crates/wasmtime/src/compiler.rs

```diff
@@ -263,10 +340,35 @@ impl<'a> CompileInputs<'a> {
             }
         }
 
-        CompileInputs { inputs }
+        // If there are any resources defined within this component, the
+        // signature for `resource.drop` is mentioned somewhere, and the
+        // wasm-to-native trampoline for `resource.drop` hasn't been created yet
+        // then insert that here. This is possibly required by destruction of
+        // resources from the embedder and otherwise won't be explicitly
+        // requested through initializers above or such.
+        if component.num_resources > 0 {
+            if let Some(sig) = types.find_resource_drop_signature() {
+                let key = CompileKey::wasm_to_native_trampoline(sig);
+                ret.resource_drop_wasm_to_native_trampoline = Some(key);
+                if !ret.input_keys.contains(&key) {
```

Comment:
It's a bit subtle and worth a comment that's not currently present, but this is doing 2 things where first it's determining if a trampoline is required and if so creating a `CompileKey` for it. Next it's detecting if that key was already compiled for some previous trampoline, and if not it adds it.

In that sense this is the "root" where `resource_drop_wasm_to_native_trampoline` is first set to `Some`, so it can't piggy-back on anything else.

That being said coupled with your below concern I'll see if I can clean this up a bit.